### PR TITLE
[DeviceMesh] Update slicing documentation to include nD and non-continuous slicing

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132632
* #132339
* __->__ #132311

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wconstab @d4l3k @c-p-i-o @tianyu-l @chauhang

### CODE REVIEW

File: torch/distributed/device_mesh.py [hunk 1]
```diff
@@ -411,31 +411,47 @@ def __getitem__(
...
-            hosts with 4 GPUs each.
-            Calling mesh["tp"] on rank 0, 1, 2, 3 would return a 1D child DeviceMesh:([0, 1, 2, 3]).
-            Calling mesh["tp"] on rank 4, 5, 6, 7 would return a 1D child DeviceMesh:([4, 5, 6, 7]).
-            Calling mesh["dp"] on rank 0, 4 would return a 1D child DeviceMesh:([0, 4]).
-            Calling mesh["dp"] on rank 1, 5 would return a 1D child DeviceMesh:([1, 5]).
-            Calling mesh["dp"] on rank 2, 6 would return a 1D child DeviceMesh:([2, 6]).
-            Calling mesh["dp"] on rank 3, 7 would return a 1D child DeviceMesh:([3, 7]).
+            The following program runs on each process/rank in an SPMD manner in a world size of 8.
+            In the first example:
+                Calling mesh_2d["tp"] on rank 0, 1, 2, 3 returns a 1D child DeviceMesh:([0, 1, 2, 3]).
```
Comment (wanchaol) [inline torch/distributed/device_mesh.py:427]: I think we should change the wording from "child DeviceMesh" to sth like "sub DeviceMesh"? A child indicates a parent somewhere, where we should avoid using this term going forward. Please rename all the comments

File: torch/distributed/device_mesh.py [hunk 2]
```diff
@@ -411,31 +411,47 @@ def __getitem__(
             self, mesh_dim_names: Union[str, Tuple[str, ...]]
         ) -> "DeviceMesh":
             """
-            Slice the current DeviceMesh based on the mesh_dim_name given to create a child
-            DeviceMesh.
+            Slice the current DeviceMesh based on the mesh_dim_names given to create a child
```
Comment (wanchaol) [inline torch/distributed/device_mesh.py:414]:
```suggestion
            Derive a sub-DeviceMesh based on the root DeviceMesh based on the ``mesh_dim_names`` provided
```

File: torch/distributed/device_mesh.py [hunk 3]
```diff
@@ -411,31 +411,47 @@ def __getitem__(
...
+            DeviceMesh. This supports 1D or nD slicing. The DeviceMesh created will have dimensions
```
Comment (wanchaol) [inline torch/distributed/device_mesh.py:415]:
```suggestion
            The DeviceMesh created consists of the dimensions and the communicators indicated by ``mesh_dim_names``
```

**Review by wanchaol — CHANGES_REQUESTED** [review]
We should change our wording given we are removing parent mesh concept, maybe this PR could be combined with the parent -> root PR together

File: torch/distributed/device_mesh.py (hunk 1 above)
Comment (wz337) [reply torch/distributed/device_mesh.py]: Removed "child DeviceMesh" and replaced with submesh instead. I am not sure if we want to expose the concept of root mesh yet, as root mesh only lives in our mesh env right now.

**Review by wanchaol — APPROVED** [review]

**Pull request closed without merging**